File: src/main/kotlin/com/io/reactivecoroutines/ReactiveCoroutinesApplication.java
package com.io.reactivecoroutines;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.data.r2dbc.repository.config.EnableR2dbcRepositories;
import reactor.blockhound.BlockHound;

import java.time.Clock;

@SpringBootApplication
@EnableR2dbcRepositories
public class ReactiveCoroutinesApplication {
    static {
        BlockHound.builder()
                // allow exception for a Kotlin reflection internal call.
                // .allowBlockingCallsInside("RandomAccessFile", "readBytes")
                .install();
    }

    @Bean
    Clock clock() {
        return Clock.systemUTC();
    }

    public static void main(String[] args) {
        SpringApplication.run(ReactiveCoroutinesApplication.class, args);
    }
}

File: src/main/kotlin/com/io/reactivecoroutines/ReactiveCoroutinesApplication.kt
package com.io.reactivecoroutines

import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.data.r2dbc.repository.config.EnableR2dbcRepositories
import reactor.blockhound.BlockHound
import reactor.blockhound.integration.BlockHoundIntegration

@SpringBootApplication
@EnableR2dbcRepositories
class ReactiveCoroutinesApplication {
    init {
        BlockHound.install(
            // WebFlux internal
            BlockHoundIntegration { builder ->
                builder.allowBlockingCallsInside(
                    "java.util.ServiceLoader\$LazyClassPathLookupIterator",
                    "parse"
                )
            },
            // Kotlin reflection
            BlockHoundIntegration { builder ->
                builder.allowBlockingCallsInside(
                    "kotlin.reflect.jvm.ReflectJvmMapping",
                    "getKotlinFunction"
                )
            },
        )
    }
}

fun main(args: Array<String>) {
    runApplication<ReactiveCoroutinesApplication>(*args)
}
